Глушаков Максим

madmax46 opened the pull request:
@coffee_and_problems

madmax46 committed:
created Unit Tests and TagsCloudVisualization proj

coffee-and-problems commented on cs/tdd.sln:
```diff
@@ -7,6 +7,12 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "BowlingGame", "BowlingGame\
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "TagsCloudVisualization", "TagsCloudVisualization\TagsCloudVisualization.csproj", "{322F5272-1669-4217-A979-BD1F10FAE50E}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Tests", "Tests", "{80C2AA2B-A45D-47F2-9437-24AC59221E8A}"
```
что это за проект?

madmax46 replied:
Это так папки выделяются в sln. Проект TagsCloudVisualizationTests содержится в папке Tests

coffee-and-problems replied:
Обычно тесты кладут рядом с проектом, на который они написаны, не создавая дополнительных папок. Но эти требования могут варьироваться в зависимости от команды

coffee-and-problems commented on cs/TagsCloudVisualizationTests/CircularCloudLayouter_Test.cs:
```diff
@@ -0,0 +1,113 @@
+    [TestCase(5, -1, TestName = "Negative height")]
+    public void PutNextRectangle_NotPositiveOrSingleSideSize_EmptyRectangle(int width, int height)
+    {
+        var rectangle = _layouter!.PutNextRectangle(new Size(0, 0));
```
тут опечатка: параметры width и height наверное должны прокидываться в Size

coffee-and-problems commented on cs/TagsCloudVisualizationTests/CircularCloudLayouter_Test.cs:
```diff
@@ -0,0 +1,113 @@
+[TestFixture]
+public class Tests
+{
+    private CircularCloudLayouter? _layouter;
```
А зачем явно сделал нуллабельным это поле? Оно же гарантировано просетится  в сетапе, тут можно убрать ? (заодно и ! ниже по коду уберутся)

madmax46 replied:
На прошлых занятиях подсмотрел эту штуку в проектах

coffee-and-problems commented on cs/TagsCloudVisualizationTests/CircularCloudLayouter_Test.cs:
```diff
@@ -0,0 +1,113 @@
+        rectangle.IsEmpty.Should().BeTrue();
+    }
+
+    [TestCase(4, 4)]
```
Почему тут именно 3 тест-кейса и именно такие? Что они проверяют? Можно ли вместо них написать один (т.е. убрать кейсы и генерировать данные в самом тесте) и просто генерировать 2 случайных числа?

madmax46 replied:
Проверяют, что переданный размер и размер полученного прямоугольника одинаковы. В целом, да, можно отказаться от тест кейсов

coffee-and-problems commented on cs/TagsCloudVisualization/Interfaces/ITagsCloudVisualization.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace TagsCloudVisualization.Interfaces;
+
+public interface ITagsCloudVisualization
```
этот файл сейчас ничего полезного не делает (как и его наследник, который просто выкидывает исключение), в коде он нигде не используется. Его логичнее было бы не добавлять в этот коммит - ведь коммит должен быть логическим завершенным кусочком. В этом коммите ты добавил построение прямоугольника на спирали, можно было закоммитить только классы, которые относятся к этому действию. А этот файлик оставить полежать до коммита, когда ты или реализуешь отрисовку облака, или тебе понадобится обратиться к методу этого интерфейса

coffee-and-problems replied:
Фишка интерфейса как раз в том, что можно везде заиспользовать именно его, а не конкретный класс, что позволит легко подменять реализации, не меняя контракт. Подробнее можно почитать в [этой статье про инверсию зависимостей](https://metanit.com/sharp/patterns/5.5.php). Сейчас вся суть существования интерфейса не используется в коде:(

madmax46 replied:
Согласен, теряется, но теряется на конкретный момент в текущей реализации, если бы мы хотели в будущем сохранять по другому, мы бы просто заменили класс, вызываемый через интерфейс.

coffee-and-problems replied:
>мы бы просто заменили класс, вызываемый через интерфейс.

Так у тебя сейчас нету вызовов через интерфейс, ты напрямую класс вызываешь)

coffee-and-problems replied:
Ну и в промышленной разработке качество кода в том числе определяется его открытостью к расширениям - если я захочу на лету подменять способы сохранения (или, например, использовать разные классы в зависимости от ОС пользователя) мне должно быть легко это сделать. Сейчас же код потребует от меня перейти в него и что-то в нем менять, потом повторно тестировать, что ничего в нем не сломалось. Это хрупкий (и не очень хороший) подход, поэтому в бизнес-коде стараются вообще все зависимости делать через интерфейсы

coffee-and-problems commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,47 @@
+
+    public Rectangle PutNextRectangle(Size rectangleSize)
+    {
+        if (rectangleSize.IsEmpty)
```
А почему тут решил возвращать такой прямоугольник, а не null или не кидать исключение, например?

madmax46 replied:
мне кажется это немного непредсказуемо кидать исключение, ведь если пользователь этого метода не обернет в дополнительный try catch, может что-то поломаться у него. Думаю возвращаться null - вполне неплохой вариант.

coffee-and-problems replied:
И все-таки почему именно такой прямоугольник?)  Это же даже не дефолтное значение для класса)

madmax46 replied:
почему не дефолтное, вполне дефолтное. Потому что возвращение пустого прямоугольника ничего не сломает, и проверка на isEmpty затратит меньше ресурсов, чем catch
![image](https://user-images.githubusercontent.com/32897932/201672368-3cf399d0-3d33-4695-87c4-07cd67281e3e.png)

coffee-and-problems replied:
Да, ты прав, Size и Rectangle - структуры, я упустила этот момент. Для них действительно дефолтное значение - это дефолтное значение всех их полей

madmax46 replied:
конечно с этим можно долго обсуждать как правильнее. Возможно самый правильный результат, это иметь метод PutNextRectangle который будет выкидывать Exception при неправильной логике, возлагая всю ответственность правильных данных на разработчика, и метод TryPutNextRectangle, который будет возвращать bool  и out Rectangle.
Это если прослеживать аналогию с generic коллекциями

coffee-and-problems commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,47 @@
+    public CircularCloudLayouter(Point center, double spiralStep = 1)
+    {
+        if (spiralStep <= 0)
+            throw new ArgumentException("Zero or negative  spiral step are not allowed");
```
маааленькая придирочка: лишний пробел закрался

coffee-and-problems commented on cs/TagsCloudVisualizationTests/CircularCloudLayouter_Test.cs:
```diff
@@ -0,0 +1,113 @@
+    [TestCase(4, 4)]
+    [TestCase(7, 3)]
+    [TestCase(5, 7)]
+    public void PutNextRectangle_TwoRectangles_NotIntersect(int width, int height)
```
Мне нравится, какие тесты ты выделил, молодец)

coffee-and-problems commented on cs/TagsCloudVisualization/ArithmeticSpiral.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace TagsCloudVisualization;
+
+public class ArithmeticSpiral
```
Потерялись тесты на этот класс

madmax46 committed:
update ArithmeticSpiral and add unit test

madmax46 committed:
update ArithmeticSpiral and add unit test

madmax46 committed:
update CircularCloudLayouter and unit tests for it

madmax46 committed:
add save rectangles to bitmap

madmax46 committed:
test

madmax46 committed:
Merge remote-tracking branch 'origin/master'

madmax46 committed:
remove not user class

madmax46 committed:
add save bitmap of tags cloud

madmax46 committed:
shift rectangle start point to its center

madmax46 committed:
generate test images of layout

madmax46 committed:
Create README.md

add readme with test layout images

coffee-and-problems commented:
![image](https://user-images.githubusercontent.com/55941390/201656734-cca164ff-4173-49a2-8a2d-f999c96b2ff7.png)

В истории что-то странное происходит, не понятно зачем делать пустой мердж этой ветки, ты же сделал все эти изменения отдельными коммитами. Можно было бы просто удалить эту ветку, раз она оказалась не нужна

coffee-and-problems commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,50 @@
+    squareRectLayout.PutNextRectangle(new Size(squareWidth, squareWidth));
+}
+
+new BitmapTagsCloudVisualization().SaveTagsCloud(squareRectLayout, "Images\\SquareRectLayout.bmp");
```
Тут не обязательно каждый раз создавать новый экземпляр BitmapTagsCloudVisualization - это просто лишний расход памяти будет. Давай вынестем поле `ITagsCloudVisualization tagsCloudVisualization = new BitmapTagsCloudVisualization();` и везде дергать его метод `tagsCloudVisualization.SaveTagsCloud(...)`. Так и избежим ненужных аллокаций и сборки мусора, и заиспользуем интерфейс, получив все плюшки инверсии зависимостей.

coffee-and-problems commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,50 @@
+
+new BitmapTagsCloudVisualization().SaveTagsCloud(horizontalRectLayout, "Images\\HorizontalRectLayout.bmp");
+
+var verticalRectLayout = new CircularCloudLayouter(new Point(0, 0), 0.01);
```
Аналогично можно вынести в переменную, классы же ничем друг от друга не отличаются. Если хочется в коде для читаемости отделить построение облаков с разными видами прямоугольников, то предлагаю использовать `#region ... #endregion`

madmax46 replied:
к сожалению в текущей реализации создание CircularCloudLayouter один раз не предусматривается, т.к. не был реализован метод Clear. Иначе все прямоугольники, что были добавлены ранее останутся

coffee-and-problems replied:
Ок, давай тогда так оставим

madmax46 committed:
use ITagsCloudVisualization interface

Use ITagsCloudVisualization interface for create visualization
of CircularCloudLayouter. Use selection code by regions.
